### ISSUE TITLE
connection reset in cursor.execute

### ISSUE THREAD

**Issue by vrbala** [uber/vertica-python#56]
Please see below

```python
import sys

sys.path.insert(0, '/home/vrbala/Downloads/vertica-python-932fe36bfb8744d903332201a0981954029f8fa2')

from vertica_python import connect

_CONNECTION_INFO = {
        'host': 'host.com',
        'port': 5433,
        'user': 'user',
        'password': 'password',
        'database': 'vertica',
        'read_timeout': 86400
        }

connection = connect(
    _CONNECTION_INFO
    )

c = connection.cursor()
c.execute("select version();")
result = c.fetchall()
print result

c.execute("select * from schema.bala_test;")
result = c.fetchall()
print result

c.execute("update schema.bala_test set x_col = 2; ")
result = c.fetchall()
print result

c.execute("commit; ")
result = c.fetchall()
print result

c.execute("insert into schema.bala_test values (1, 1); ")
result = c.fetchall()
print result

c.execute("commit; ")
result = c.fetchall()
print result

c.execute("select * from schema.bala_test;")
result = c.fetchall()
print result

connection.close()
```

When I run the script,

% python vertica-test.py
[[u'Vertica Analytic Database v7.1.1-11']]
[[Decimal('1'), u'1'], [Decimal('1'), u'1']]
[[2]]
[]
[[1]]
[]
[[Decimal('1'), u'1'], [Decimal('1'), u'1']]

The test table had 2 lines originally [[1,1],[1,1]]. If you notice even after the update and insert queries, nothing has changed in the tables. 

Analysis shows that the reset_connection call in cursor.execute causes every execute call to behave like a transaction and essentially a new connection. This behavior is not correct and breaks code that relies on a connection object representing a single connection & transaction unless user chooses to reboot it.

**Comment by zer0n** [comment]
Please review the fix and test if it indeed fixes the issue: https://github.com/uber/vertica-python/commit/a04d4ec5fc5ab96256f5d243ed62d6526ed705e7

**Comment by zer0n** [comment]
+ @jberka

**Comment by zer0n** [comment]
I haven't tested and am not sure if the issue is due to flushed connection or due to incomplete processing for queries. Regardless, the changes would make the code more robust even if it doesn't address the issue reported here.

**Comment by vrbala** [comment]
I can test this on Monday. Meanwhile I have posted my comments and some questions based on the observations.

**Comment by jberka** [comment]
Wall of text incoming...

@zer0n Thanks for the fix; I think this should properly handle the CommandComplete (I suspect that on a COMMIT, the message sequence is CommandComplete followed by a ReadyForQuery, which should now get picked up). There is some additional data (CommandComplete messages should return a row count and a tag, similar to ReadyForQuery returning the transaction status) that could be added to the connection, but I'm not sure how useful it is.

The current version of the library looks a bit different from the original, but I suspect the ```fetchone()``` function may also need an ```else: self.connection.process_message(...)``` call (in order to process the final ReadyForQuery message after DataRows have been returned).

Regarding the open transaction issue described above, I don't think your fix will solve that, since the cursor's ```last_execution``` isn't getting unset anywhere (so the subsequent ```execute()``` calls still reset the connection). I'm not sure open transactions are necessarily a use case that should be supported, though, since it seems like a good way to get in trouble (process some rows from somewhere else -> buffer the processed rows and discard the original rows -> execute an INSERT without a commit -> Vertica goes down -> trouble).

@vrbala What's the use case for leaving an INSERT open in a transaction here? Also, could you enable debug logging to see if you can pull the full message log from when your script runs? It would be really helpful to see what Vertica is sending back in each case.

**Comment by vrbala** [comment]
>> I'm not sure open transactions are necessarily a use case that should be supported, though, since it seems like a good way to get in trouble (process some rows from somewhere else -> buffer the processed rows and discard the original rows -> execute an INSERT without a commit -> Vertica goes down -> trouble).

IMHO, open transactions should be supported as this gives the user an option to fire an insert/update and commit/rollback the changes later. I know this pattern is more common in OLTP than OLAP, but when data corrections happen in OLAP I think this gives more flexibility than forcing the user to combine a commit in every insert/update. And, backward compatibility - I know that this breaks code that was running fine till vertica-python 0.2.3. 

See  below for the debug information for the script above - 

```
% python vertica-test.py
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7ffff0a74d90>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee3b0fd0>


version
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7ffff0a74d90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7ffff0a74d90>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b0d0>
<vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee3b0f90>
[[u'Vertica Analytic Database v7.1.1-11']]


bala_test
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee39b0d0>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee39b090>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b050>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b050>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b0d0>
<vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee3b0f90>
[[Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('2'), u'1'], [Decimal('2'), u'1'], [Decimal('2'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1']]


update set x=2
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee39b0d0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee39b0d0>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b110>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b110>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b090>
<vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee3b0f90>
[[13]]


commit
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee39b150>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee39b150>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee39b090>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.notice_response.NoticeResponse object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.notice_response.NoticeResponse object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee39b090>
<vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee39b090>
[]


insert
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b150>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee3b0fd0>
[[1]]


commit after insert
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee39b050>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee39b050>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee39b090>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee39b090>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.notice_response.NoticeResponse object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.notice_response.NoticeResponse object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee39b090>
<vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee39b090>
[]


select after insert
=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.startup.Startup object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0fd0>
=> %s <vertica_python.vertica.messages.frontend_messages.password.Password object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.authentication.Authentication object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee39b050>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.backend_key_data.BackendKeyData object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.ready_for_query.ReadyForQuery object at 0x7fffee3b0f90>
=> %s <vertica_python.vertica.messages.frontend_messages.query.Query object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.row_description.RowDescription object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<vertica_python.vertica.messages.backend_messages.parameter_status.ParameterStatus object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0f90>
<= %s <vertica_python.vertica.messages.backend_messages.data_row.DataRow object at 0x7fffee3b0fd0>
<= %s <vertica_python.vertica.messages.backend_messages.command_complete.CommandComplete object at 0x7fffee3b0f90>
[[Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('2'), u'1'], [Decimal('2'), u'1'], [Decimal('2'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1'], [Decimal('1'), u'1']]

=> %s <vertica_python.vertica.messages.frontend_messages.terminate.Terminate object at 0x7fffee3b0fd0>
```

**Comment by zer0n** [comment]
@jberka Good points. I've added calling connection.process_message if message is not a data row in fetchone() to the PR.

@vrbala If we clear the message buffer in execute instead of reset the connection, that would solve the problem. Would you want to go ahead with a PR?

**Comment by jberka** [comment]
@vrbala Thanks for the message log; it definitely looks like a ReadyForQuery is being missed in some cases.

@zer0n Thanks. I finally downloaded the Vertica VM, so I may try and make a few other fixes/add tests/etc.

**Comment by vrbala** [comment]
@zer0n No, please go ahead with the change as you know the context much better. 

@jberka Is vertica VM publicly downloadable? I tried to stand up a community edition on my laptop but couldn't get it up working (mostly configuration & certificate issues).

**Comment by jberka** [comment]
Yup, if you download VirtualBox (https://www.virtualbox.org/) and the "VMWare vSphere 4.0 (OVF)" virtual machine from https://my.vertica.com/download-community-edition/, it only takes a few minutes to get a database set up (via admintools).

**Comment by vrbala** [comment]
Awesome, thanks!

@zer0n I fixed one more bug in my tree. I will send a PR for clearing message buffer as well as the new issue tonight PDT. I will base it off of commit 6b980016bc461bc2a3f58e86614aa3a2479c147e.

**Comment by Bennyelg** [comment]
I tried to dig into this issue but couldn't understand why it's happening. maybe its a transnational problem ?

**Comment by alexjikim** [comment]
i think this should be fixed as of 0.4.5. (https://github.com/uber/vertica-python/pull/63)
Please try upgrading and reopen this if you still see issues.
